SwiftSyntaxDevUtils/Sources/swift-syntax-dev-utils/common/Builder.swift

```diff
@@ -0,0 +1,102 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+protocol Builder {
+  var arguments: BuilderArguments { get }
+}
+
+extension Builder {
+  func buildProduct(productName: String) throws {
+    logSection("Building product " + productName)
+    try build(packageDir: Paths.packageDir, productName: productName)
+  }
+
+  func buildExample(exampleName: String) throws {
+    logSection("Building example " + exampleName)
+    try build(packageDir: Paths.examplesDir, productName: exampleName)
+  }
+
+  @discardableResult
+  func invokeSwiftPM(
+    action: String,
+    packageDir: URL,
+    additionalArguments: [String],
+    additionalEnvironment: [String: String]
+  ) throws -> String {
+    var args = [action]
+    args += ["--package-path", packageDir.path]
+
+    if let buildDir = arguments.buildDir?.path {
+      args += ["--scratch-path", buildDir]
+    }
+
+    #if !canImport(Darwin)
+    args += ["--enable-test-discovery"]
+    #endif
+
+    if arguments.release {
+      args += ["--configuration", "release"]
+    }
+
+    if let multirootDataFile = arguments.multirootDataFile?.path {
+      args += ["--multiroot-data-file", multirootDataFile]
+    }
+
+    if arguments.disableSandbox {
+      args += ["--disable-sandbox"]
+    }
+
+    if arguments.verbose {
+      args += ["--verbose"]
+    }
+
+    args += additionalArguments
+
+    let processRunner = ProcessRunner(
+      executableURL: arguments.toolchain.appendingPathComponent("bin").appendingPathComponent("swift"),
+      arguments: args,
+      additionalEnvironment: additionalEnvironment
+    )
+
+    let result = try processRunner.run()
+
+    return result.stdout
+  }
+
+  private func build(packageDir: URL, productName: String) throws {
+    let args = ["--product", productName]
+
+    var additionalEnvironment: [String: String] = [:]
+    additionalEnvironment["SWIFT_BUILD_SCRIPT_ENVIRONMENT"] = "1"
+
+    if arguments.enableRawsyntaxValidation {
+      additionalEnvironment["SWIFTSYNTAX_ENABLE_RAWSYNTAX_VALIDATION"] = "1"
+    }
+
+    if arguments.enableTestFuzzing {
+      additionalEnvironment["SWIFTPARSER_ENABLE_ALTERNATE_TOKEN_INTROSPECTION"] = "1"
+    }
+
+    // Tell other projects in the unified build to use local dependencies
+    additionalEnvironment["SWIFTCI_USE_LOCAL_DEPS"] = "1"
+    additionalEnvironment["SWIFT_SYNTAX_PARSER_LIB_SEARCH_PATH"] = arguments.toolchain.appendingPathComponent("/lib/swift/macos").path
```

Comment:
This should be calling `appendingPathComponent` multiple times for every component.